src/lambda/cached.rs

```diff
@@ -0,0 +1,97 @@
+use core::future::Future;
+use std::num::NonZeroU64;
+use std::pin::Pin;
+use std::sync::{Arc, RwLock};
+
+use anyhow::Result;
+use async_graphql_value::ConstValue;
+
+use super::{Concurrent, Eval, EvaluationContext, Expression, ResolverContextLike, IO};
+use crate::lambda::has_io::HasIO;
+
+pub trait CacheKey<Ctx> {
+    fn cache_key(&self, ctx: &Ctx) -> u64;
+}
+
+#[derive(Clone, Debug)]
+pub enum Cached {
+    IOCache(IOCache),
+    NonIOCache(NonIOCache),
+}
+
+#[derive(Clone, Debug)]
+pub struct IOCache {
+    max_age: NonZeroU64,
+    expr: Box<Expression>,
+}
+
+#[derive(Clone, Debug)]
+pub struct NonIOCache {
+    data: Arc<RwLock<Option<ConstValue>>>,
+    expr: Box<Expression>,
+}
+
+impl Cached {
+    pub fn new(max_age: NonZeroU64, expr: Expression) -> Self {
```

Comment:
```suggestion
    pub fn wrap(max_age: NonZeroU64, expr: Expression) -> Expression {
```

Move the IO check outside of Cached::wrap. You are anyways check for one condition which is — Does the field have a cache operator, add this second check about it the field has an IO operator. Essentially we are performing a "Cachability" test.

Reply:
Alternative way could be to move all the logic into wrap.
- Check if the resolver can be cached
- Check if there is an IO operation
- Identify the cache key
- Return an Expression that has access to the cache key generator function, the original expression and the TTL.